### DISCUSSION TITLE
Installation Issues: Failed to build chroma-hnswlib

### DISCUSSION THREAD

**Discussion by isudoajl** [danielmiessler/Fabric#765] [Q&A] [answered]
`Successfully built fabric
Failed to build chroma-hnswlib

PIP STDERR
----------
WARNING: embedchain 0.1.118 does not provide the extra 'github'
WARNING: embedchain 0.1.118 does not provide the extra 'youtube'
WARNING: embedchain 0.1.117 does not provide the extra 'github'
WARNING: embedchain 0.1.117 does not provide the extra 'youtube'
WARNING: typer 0.12.3 does not provide the extra 'all'
WARNING: typer 0.12.2 does not provide the extra 'all'
WARNING: typer 0.12.1 does not provide the extra 'all'
WARNING: typer 0.12.0 does not provide the extra 'all'
WARNING: typer 0.12.3 does not provide the extra 'all'
WARNING: typer 0.12.2 does not provide the extra 'all'
WARNING: typer 0.12.1 does not provide the extra 'all'
WARNING: typer 0.12.0 does not provide the extra 'all'
  error: subprocess-exited-with-error
  
  × Building wheel for chroma-hnswlib (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [21 lines of output]
      running bdist_wheel
      running build
      running build_ext
      creating tmp
      gcc -fno-strict-overflow -Wsign-compare -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -fexceptions -fcf-protection -fexceptions -fcf-protection -fexceptions -fcf-protection -fPIC -I/home/isudoajl/.local/share/pipx/venvs/fabric/include -I/usr/include/python3.12 -c /tmp/tmp6fxv6bj4.cpp -o tmp/tmp6fxv6bj4.o -std=c++14
      gcc -fno-strict-overflow -Wsign-compare -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -fexceptions -fcf-protection -fexceptions -fcf-protection -fexceptions -fcf-protection -fPIC -I/home/isudoajl/.local/share/pipx/venvs/fabric/include -I/usr/include/python3.12 -c /tmp/tmpn1ca_jq8.cpp -o tmp/tmpn1ca_jq8.o -fvisibility=hidden
      building 'hnswlib' extension
      creating build
      creating build/temp.linux-x86_64-cpython-312
      creating build/temp.linux-x86_64-cpython-312/python_bindings
      gcc -fno-strict-overflow -Wsign-compare -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -fexceptions -fcf-protection -fexceptions -fcf-protection -fexceptions -fcf-protection -fPIC -I/tmp/pip-build-env-9kxbtv2v/overlay/lib/python3.12/site-packages/pybind11/include -I/tmp/pip-build-env-9kxbtv2v/overlay/lib64/python3.12/site-packages/numpy/_core/include -I./hnswlib/ -I/home/isudoajl/.local/share/pipx/venvs/fabric/include -I/usr/include/python3.12 -c ./python_bindings/bindings.cpp -o build/temp.linux-x86_64-cpython-312/./python_bindings/bindings.o -O3 -march=native -fopenmp -DVERSION_INFO=\"0.7.3\" -std=c++14 -fvisibility=hidden
      In file included from /tmp/pip-build-env-9kxbtv2v/overlay/lib/python3.12/site-packages/pybind11/include/pybind11/detail/../attr.h:13,
                       from /tmp/pip-build-env-9kxbtv2v/overlay/lib/python3.12/site-packages/pybind11/include/pybind11/detail/class.h:12,
                       from /tmp/pip-build-env-9kxbtv2v/overlay/lib/python3.12/site-packages/pybind11/include/pybind11/pybind11.h:13,
                       from /tmp/pip-build-env-9kxbtv2v/overlay/lib/python3.12/site-packages/pybind11/include/pybind11/functional.h:12,
                       from ./python_bindings/bindings.cpp:2:
      /tmp/pip-build-env-9kxbtv2v/overlay/lib/python3.12/site-packages/pybind11/include/pybind11/detail/../detail/common.h:274:10: fatal error: Python.h: No such file or directory
        274 | #include <Python.h>
            |          ^~~~~~~~~~
      compilation terminated.
      error: command '/usr/bin/gcc' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for chroma-hnswlib
ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (chroma-hnswlib)
`
**Distro: Fedora 40 
Python version: Python 3.12.4**

I already try everything that I found here and nothing works:
With and without using python venv
With and without https://install.python-poetry.org/

I got always the same error :/

Here https://github.com/danielmiessler/fabric/discussions/657 he say that resolve the problem:

This was resolved by getting pip dependencies installed and running a python environment from cd ~

I try it but I got the same error.

**Comment by isudoajl** [comment, marked as answer]
The solution was install the package **python3.12-devel and gcc-c++**:
`sudo dnf install python<here_the_version>-devel`
Example:
`sudo dnf install python3.12-devel`
`sudo dnf install gcc-c++`

In Debian-based distributions **python3.12-dev**
`sudo dnf apt-get python<here_the_version>-dev`
Example:
`sudo apt-get install python3.12-dev`

**Reply by username-username-username** [reply]
I am not 100% expert but these steps didn't work for me on Fedora 40 but using this `sudo dnf install gcc-c++` after doing `sudo dnf install python3.12-devel` and then running the normal install process worked for me. Does that sound plausible to everyone or was that a fluke thing? I kept getting errors 

```shell
 pip failed to build package:
    chroma-hnswlib

Some possibly relevant errors from pip install:
    error: subprocess-exited-with-error
    gcc: fatal error: cannot execute ‘cc1plus’: posix_spawnp: No such file or directory
    RuntimeError: Unsupported compiler -- at least C++11 support is needed!
    ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (chroma-hnswlib)

Error installing fabric from spec '[file-location-here]/fabric'.
```

But after the steps above then running normal install process worked like normal. This was only on Fedora 40 by the way I previsously was using this on Ubutntu 24.04LTS and the normal install procedures worked.
